### PR TITLE
Object.prototype.toString: split Symbol.toStringTag tests into separate files

### PR DESCRIPTION
**Pull request by ljharb**
See https://github.com/tc39/test262/pull/3855#issuecomment-1593543655

I also added some test coverage that was previously missing.

### CODE REVIEW

File: test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  15. Let tag be ? Get(O, @@toStringTag).
+  16. If Type(tag) is not String, set tag to builtinTag.
+  17. Return the string-concatenation of "[object ", tag, and "]".
+features: [Symbol.toStringTag, Symbol.iterator, generators, iterator-helpers]
+---*/
+
+var toString = Object.prototype.toString;
+
+var genFn = function* () {};
+assert.sameValue(toString.call(gen), '[object GeneratorFunction]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js:20]: this line was added

File: test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+var gen = genFn();
+assert.sameValue(toString.call(gen), '[object Generator]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js:23]: this line was added

File: test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+var genProto = Object.getPrototypeOf(gen);
+Object.defineProperty(genProto, Symbol.toStringTag, {
+  get: function() { return {}; },
+});
+assert.sameValue(toString.call(gen), '[object Iterator]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-generators-builtin.js:29]: this line was _corrected_ - it was "Object" but it should be "Iterator"

File: test/built-ins/Object/prototype/toString/symbol-tag-map-builtin.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+delete Map.prototype[Symbol.toStringTag];
+assert.sameValue(toString.call(map), '[object Object]');
+
+var mapIter = map[Symbol.iterator]();
+var mapIterProto = Object.getPrototypeOf(mapIter);
+assert.sameValue(toString.call(mapIter), '[object Map Iterator]');
+Object.defineProperty(mapIterProto, Symbol.toStringTag, {
+  get: function() { return new String('ShouldNotBeUnwrapped'); },
+});
+assert.sameValue(toString.call(mapIter), '[object Iterator]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-map-builtin.js:29]: this line was corrected - it was Object but should be Iterator

File: test/built-ins/Object/prototype/toString/symbol-tag-promise-builtin.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  15. Let tag be ? Get(O, @@toStringTag).
+  16. If Type(tag) is not String, set tag to builtinTag.
+  17. Return the string-concatenation of "[object ", tag, and "]".
+features: [Symbol.toStringTag, Symbol.iterator, Promise]
+---*/
+
+var toString = Object.prototype.toString;
+
+var promise = new Promise(function () {});
+assert.sameValue(toString.call(promise), '[object Promise]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-promise-builtin.js:20]: this line was added

File: test/built-ins/Object/prototype/toString/symbol-tag-set-builtin.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+delete Set.prototype[Symbol.toStringTag];
+assert.sameValue(toString.call(map), '[object Object]');
+
+var mapIter = set[Symbol.iterator]();
+var mapIterProto = Object.getPrototypeOf(mapIter);
+assert.sameValue(toString.call(mapIter), '[object Set Iterator]');
+Object.defineProperty(mapIterProto, Symbol.toStringTag, {
+  get: function() { return new String('ShouldNotBeUnwrapped'); },
+});
+assert.sameValue(toString.call(mapIter), '[object Iterator]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-set-builtin.js:29]: this line was corrected from Object to Iterator

File: test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+var toString = Object.prototype.toString;
+
+var arrIter = [][Symbol.iterator]();
+var arrIterProto = Object.getPrototypeOf(arrIter);
+
+assert.sameValue(toString.call(arrIter), '[object Array Iterator]');
+
+Object.defineProperty(arrIterProto, Symbol.toStringTag, {value: null});
+assert.sameValue(toString.call(arrIter), '[object Iterator]');
```
Comment (ljharb) [inline test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js:25]: this line was corrected from Object to Iterator

Comment (syg) [reply test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js:25]: No, the original test was correct. The line right above this one redefines %IteratorPrototype%'s @@toStringTag to be `null`, so it should be `[object Object]`. Please also double check the other changed tests.

Comment (ljharb) [reply test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js:25]: It redefines ArrayIteratorPrototype, not IteratorPrototype, i believe.

Comment (syg) [reply test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js:25]: Quite right. Sorry, my reasoning was wrong but the original test is still correct.

`Object.defineProperty(arrIterProto, Symbol.toStringTag, {value: null});` sets the value of @@toStringTag to `null`. Steps 14 and 15 of [Object.prototype.toString](https://tc39.es/ecma262/#sec-object.prototype.tostring) are:

15. Let tag be ? [Get](https://tc39.es/ecma262/#sec-get-o-p)(O, [@@toStringTag](https://tc39.es/ecma262/#sec-well-known-symbols)).
16. If tag [is not a String](https://tc39.es/ecma262/#sec-ecmascript-language-types-string-type), set tag to builtinTag.

So if it finds `null` on `arrIterProto`, finds it unusable, and returns `[object Object]`.

Comment (ljharb) [reply test/built-ins/Object/prototype/toString/symbol-tag-array-builtin.js:28]: Updated - now it verifies its Object when unusable, then deletes it, and then verifies it’s Iterator.

**Comment by ljharb** [comment]
@syg ping on taking another look here

**Comment by ptomato** [comment]
I agreed to review this one. Sorry it has taken me so long to get to it. I expect to have time this week.

**Review by ptomato — APPROVED** [review]
Thanks, looks good. Sorry for the long wait.

**Pull request merged**